### PR TITLE
add django model inference

### PR DESCRIPTION
**Pull request by lime-green**
Tested `count_calls_on_model` on THM and it catches all the cases we were catching of `save()` from before plus a few other (valid) cases 

Provides the following APIs:

```
from codewatch import count_calls_on_model

# If any inferred usage of dangerous_method() against a DjangoUser object occurs, then
# stats.get(STATS_NAMESPACE) will contain a dictionary keyed by the file name 
# and value being the number of calls of 'dangerous_method' in that file
count_calls_on_model(
    STATS_NAMESPACE,
    'app.lobby.models.User',
    'dangerous_method',
)

@assertion(stats_namespaces=[STATS_NAMESPACE]):
def has_expected_dangerous_calls(stats)
  assert stats == {'app/lobby/views.py': 3, 'app/lobby/util.py': 1}, 'Mismatch'
```

and

```
from codewatch import get_inference_for_model

inference = get_inference_for_model('app.lobby.models.User')

# can now be used in view functions where you would expect to need to infer User objects
@visit(..., inferences=[inference])
def my_visitor(...)
  pass
```

### CODE REVIEW

File: codewatch/helpers/visitors.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+            return call_node
+
+        record_stats(stats, rel_file_path)
+
+        return call_node
+
+    NodeVisitorMaster.register_visitor(Call, visit_call, inferences=inferences)
+
+
+def count_calls_on_model(stats_namespace, model_qname, method_name):
```
Comment (lime-green) [inline codewatch/helpers/visitors.py]: This function was the only new code in this file

File: codewatch/helpers/inference.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    'get_or_create',
+    'update_or_create',
+) + DJANGO_MANAGER_METHODS_LIST
+
+
+CODEWATCH_MODEL_INFERENCE_KEY = '_codewatch_inferred_model_klass_obj'
+
+
+def get_inference_for_model(model_qname):
+    def _did_we_infer(node):
```
Comment (cabiad) [inline codewatch/helpers/inference.py]: Name is a little surprising given the tuple returned. Maybe:

```suggestion
     def _get_inference_details(node):
```

?

File: codewatch/helpers/inference.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+            eg: all_users.first()
+        """
+        if (
+            not hasattr(node.func, 'attrname')
+            or node.func.attrname not in DJANGO_MANAGER_METHODS
+        ):
+            return False
+
+        if not hasattr(node.func, 'expr'):
+            return
```
Comment (cabiad) [inline codewatch/helpers/inference.py]: ?

```suggestion
            return False
```

File: codewatch/helpers/visitor.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+        if not found_matching_inferred_qname:
+            return call_node
+
+        record_stats(stats, rel_file_path)
+        return call_node
+
+    NodeVisitorMaster.register_visitor(Call, visit_call, inferences=inferences)
+
+
+def count_calls_on_model(stats_namespace, model_qname, method_name):
```
Comment (cabiad) [inline codewatch/helpers/visitor.py]: maybe the name should include the word `django`?

**Comment by cabiad** [comment]
Lgtm

**Pull request merged**